server/src/main/java/org/opensearch/action/search/SearchRequestListenerManager.java

```diff
@@ -0,0 +1,124 @@
+        searchRequestListenersList = new ArrayList<>();
+    }
+
+    /**
+     * Add a {@link SearchRequestOperationsListener} to the searchRequestListenersList,
+     * which will be executed during each search request.
+     *
+     * @param listener A SearchRequestOperationsListener object to add.
+     * @throws IllegalArgumentException if the input listener is null or already exists in the list.
+     */
+    public void addListener(SearchRequestOperationsListener listener) {
+        if (listener == null) {
+            throw new IllegalArgumentException("listener must not be null");
+        }
+        if (searchRequestListenersList.contains(listener)) {
+            throw new IllegalArgumentException("listener already added");
+        }
+        searchRequestListenersList.add(listener);
+    }
+
+    /**
+     * Remove a {@link SearchRequestOperationsListener} from the searchRequestListenersList,
+     *
+     * @param listener A SearchRequestOperationsListener object to remove.
+     * @throws IllegalArgumentException if the input listener is null or already exists in the list.
+     */
+    public void removeListener(SearchRequestOperationsListener listener) {
+        if (listener == null) {
+            throw new IllegalArgumentException("listener must not be null");
+        }
+        if (!searchRequestListenersList.contains(listener)) {
+            throw new IllegalArgumentException("listener does not exist in the listeners list");
+        }
+        searchRequestListenersList.remove(listener);
+    }
+
+    /**
+     * Get searchRequestListenersList,
+     *
+     * @return List of SearchRequestOperationsListener
+     * @throws IllegalArgumentException if the input listener is null or already exists in the list.
+     */
+    public List<SearchRequestOperationsListener> getListeners() {
+        return searchRequestListenersList;
+    }
+
+
+    /**
+     * Create the {@link SearchRequestOperationsListener.CompositeListener}
+     * with the all listeners enabled at cluster-level and request-level.
+     *
+     * @param searchRequest The SearchRequest object. SearchRequestListenerManager will decide which request-level listeners to add based on states/flags of the request
+     * @param logger Logger to be attached to the {@link SearchRequestOperationsListener.CompositeListener}
+     * @param perRequestListeners the per-request listeners that can be optionally added to the returned CompositeListener list.
+     * @return SearchRequestOperationsListener.CompositeListener
+     */
+    public SearchRequestOperationsListener.CompositeListener buildCompositeListener(
+        SearchRequest searchRequest,
+        Logger logger,
+        SearchRequestOperationsListener... perRequestListeners
+    ) {
+        final List<SearchRequestOperationsListener> searchListenersList = new ArrayList<>(searchRequestListenersList);
+
+        Arrays.stream(perRequestListeners).parallel().forEach((listener) -> {
```

Comment:
The only hesitation I had was where to put the "add timeProvider" logic. I wanted to remove [this whole `createSearchListenerList` function](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/search/TransportSearchAction.java#L1247) in the TransportSearchAction.java. However it was tricky to deal with that time provider since that object would [be used later](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/search/TransportSearchAction.java#L522) in the downstream code. That's the reason I'm still keeping the logic to create a timeProvider in TransportSearchAction.java, but the logic to add it to the listeners list was moved to the listener manager.

Reply:
You definitely don't want to do the `parallel()` call here. 

In particular, it's going to modify `searchListenersList` across multiple threads (which is likely to cause a `ConcurrentModificationException`).

Reply:
In fact, the`searchListenersList` is not modified in this code block, and this `parallel` will only make the workflow to decide which "perRequestListener" I want to add to the returned listeners list . In this case would it be safe to use `parallel` here?